Regex search

Alkarex opened the pull request:
fix https://github.com/FreshRSS/FreshRSS/issues/3549

Support regex filters and regex search (working with PostgreSQL and MySQL and SQLite).

Works like any regular expression. Must be enclosed in `/ /` 

Currently supporting the prefix `intitle:` or no prefix at all (will add `author:` etc. soon)

Case sensitive by default, but can be made case-insensitive with `i` modifier like `/Alice/i`

Supports multiline mode with `m` modifier like `/^Alice/m`

Example of filter to keep only the entries, which title ends with `.0` from https://github.com/FreshRSS/FreshRSS/releases.atom , marking as read the entries not matching this filter:
![image](https://github.com/user-attachments/assets/ddffc799-aeca-4faf-860e-7978bac26e05)

![image](https://github.com/user-attachments/assets/b5e4f5d2-b9d9-4a71-ac08-70bec20a7302)

Requires PHP 8.0+ due to use of new function `str_contains`, as planned requirement for FreshRSS 1.25.0+ https://github.com/FreshRSS/FreshRSS/pull/6711

Alkarex committed:
Regex search
fix https://github.com/FreshRSS/FreshRSS/issues/3549

Alkarex committed:
Fix PHPStan

Alkarex committed:
Fix escape

Alkarex committed:
Fix ungreedy

Alkarex committed:
Initial support for regex search in PostgreSQL and MySQL

Alkarex committed:
Improvements, support MySQL

Alkarex committed:
Fix multiline

Alkarex committed:
Add support for SQLite

Alkarex committed:
A few tests

Alkarex committed:
Merge branch 'edge' into regex

Alkarex committed:
Merge branch 'edge' into regex

mtalexan commented:
I'm running this branch in a custom docker-compose build based on the `Docker-Alpine`. I see you pretty much only have `intitle:` and the prefix-less (context) matching implemented so far so I'm focusing on testing regex features and combos of the regex feature with other pre-existing filter types.

---

I'd recommend being explicit about what type of regex patterns you support.  The most common standard types are Posix (very limited), Extended, and Perl, though there are lots of variations like Java, RE2 (Golang), etc that slightly tweak the syntax, usually primarily only of more advanced features. Maybe if you add a link in the documentation to the regex engine's definition that would be helpful as well. 

Working:
- `intitle` with regex
- `intitle` with case insensitive regex
- content with regex
- content with case insensitive regex
- `intitle` with regex and content with non-regex
- `intitle` with case insensitive regex and non-regex
- `intitle` with regex ORed with `intitle` regex
- `intitle` without regex
- content without regex
- intitle regex and tag (non-regex) pattern
- character exclusion regex patterns (i.e. `[^a-z]`)
- Wildcard match patterns (`.*` between words)
- Start of boundary patterns (`/^Israel/`)
- One-or-more wildcard match patterns (`[^a-z]+`)
- End of boundary match patterns (`routing\.$`)
- Alternatives matching, though see the issues with grouping. `messages?( routing|$)`

Not working:
- grouping in regex patterns (sort of, see below)
- multi-line mode (see below)
- (user error?) explicit `and` between patterns. i.e. `intitle:/election/i and intitle:/year/i`
- (pre-existing?) `inurl` (non-regex) can't match anything (i.e. `inurl:'p88#')
- (pre-existing?) matches are applied to the raw text, not rendered HTML. i.e. `'hit 100,000 downloads'` where the `100,000 downloads` part is a hyperlink fails to match.
- (not supported?) Look ahead patterns (`/election (?!year)/i`)

## Bugs

### Pattern Parenthesis

It's common to use parentheses in regex patterns to form conditional groups.  For example: `intitle:/messages( routing|$)/i`  to match either `messages` at the end of a line, or `messages routing` anywhere in the title.  

The parenthesis currently require manual escaping however, they're being captured by the filter logic outside the regex.  This also appears to be an existing bug in how the non-regex filters work as well, but it's much more significant for regex patterns because parentheses are used so frequently in them.

For example, this fails to match anything `intitle:/messages( routing|$)/i`, but this `intitle:/messages\( routing|$\)/i` matches two different articles, one with `messages routing` in the middle of the title, and another that ends with `messages`.  I also verified that literal parentheses aren't being matched by unescaped parentheses in the pattern, `intitle:/(Zachary/i` fails to match a title containing `(Zachary Cohen/CNN)` in it. In fact, matching a literal parenthesis requires escaping it from both the filter engine (`\(`) and the regex pattern (`[(]`), resulting in needing to specify `[\(]` and `[\)]`.

I did verify that this same parenthesis bug affects quoted non-regex patterns though too.  Doing `infile:'(Zachary'` is also failing to match the title with `(Zachary Cohen/CNN)` in it, and I have to do `infile:'\(Zachary'` to get it to match.  It's just less severe there because literal parentheses are much less common than needing to use a parenthesis in a regex pattern.


### Multi-line mode

I'm not sure how to test multi-line mode exactly.  It works fine if I put the `m` on an otherwise working case sensitive or insensitive pattern, but it doesn't add any functionality to do so.  With text like this in the middle of the content of an item:
```
Better than Signal? Looks promising.
A few bugs and UX issues but great foundation. Love that it’s public domain.
```

The pattern `/Looks promising\./` works, as does `/looks promising\./i` and I can put an `m` on the end of either and it still works (`/Looks promising\./m` or `/looks promising\./im`).   
But if I try to match across the lines: `/Looks promising\..*A few bugs and UX issues/m` it fails to match.  Or if I try to explicitly match the newlines it also doesn't work: `/Looks promising\.(\r)?\n(\r)?A few/im` (though maybe there's some escaping issues here).  Here's no way to insert a literal newline in the ad-hoc search box that I've been doing most of my testing with that I'm aware of, so I'd have to have some way to match past a line ending.  

I suspect you want `.` to match line endings as well, and probably need to set a special option for the regex engine to let `.` match those end-of-line terminators when multi-line mode is enabled.  For some reason that's frequently something regex engines require you to separately enable from multi-line mode, even though they'd really only ever expect to be used together.

Alkarex commented:
Hello @mtalexan and thanks for testing 👍🏻 
Quick answer already:
* there is no explicit `AND` operator; just use a space instead.
* The regex engine is not fully consistent and depends on which database you are using:
    * [PHP preg_match](https://php.net/function.preg-match) for auto-mark-as-read and auto-favourite, as well as for searching with an SQLite database
    * [For PostgreSQL search](https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-POSIX-REGEXP)
    * [For MySQL search](https://dev.mysql.com/doc/refman/8.4/en/regexp.html#function_regexp-like)
* Searches are indeed applied to the raw HTML content
* The multiline mode allows searching for instance for a text at the start or end of a line in the raw HTML content

Alkarex committed:
Merge branch 'edge' into regex

mtalexan commented:
Excellent, thanks for the clarification.  For reference, I'm testing with PostgreSQL.

- Lookahead is working, it just needed the parentheses escaped.
- Literal parentheses can be escaped with `\\(`, i.e. an escaped `\\`, but for some reason a single `\` not preceeding a `(` never needs to be escaped.
- Multi-line matching is working, it's just usually hard to find a use for since multi-line content usually has HTML or something included that's harder to match against. The `.` is non-newline-sensitive like you'd expect/want.
- Literal `/` in a pattern can successfully be escaped (`\/`)

It seems only parentheses have weird escaping rules.  Pattern escape characters like `\d` don't need the `\` part escaped. Only a `(` or `)` in the regex needs to be escaped, or a `\` before a `(` or `)` that is supposed to be in the regex pattern (i.e. `\\(` and `\\)` -> `\(` and `\)` in the pattern).  That's incredibly strange behavior because it means `\` isn't a consistent escape character like almost every other parser ever written, instead there's weird multi-character matching that must be going on instead of nested interpretation(?).  

The super strange behavior with parenthesis and escaping definitely need to be documented. No one will ever figure out how to use it without a lot of trial-and-error, and it's such a standard feature of regex patterns to use parentheses you can guarantee almost everyone will encounter the issue. I'd suggest dropping the mention in the documentation on filtering along with the links to the regex syntaxes supported by the different databases. 

---

Is there anything else you'd like me to test?  
The existing non-regex patterns still work the same, a solid sample of the range of all supported regex features work for both case-insensitive and multi-line mode, and can be used in combination with other regex and/or non-regex patterns.

It looks like the same code path is used for all cases of filters being used, so I assume there's not much point in re-testing with per-feed filters, label filters, or saved user filters.  Correct?

mtalexan commented:
Oh, forgot to mention I'd tested negation of a regex filter in my original set of tests.  It's working fine too.

mtalexan commented:
As I work on using the regexes more, I'm finding more and more cases where it's not clear if there was a regex error, or if the pattern is matching unexpected things.  For example, if I use the word-boundary escape character (`\y`), it doesn't seem to be working and the only way I can determine that is because the same top 10 things appear in my view as if I had no search filter.  

This for example doesn't work: `intitle:/\yart/i, and it causes 10 items to appear in the view that don't match anything close to what I'm looking for (the case-insensitive prefix "art", making use of the `\y` constraint escape for a word boundary).  Recognizing the top list of items are unchanged despite the whole list blanking and re-loading, and none of them match what I'm trying to search for isn't exactly easy to determine.

---

Related: it appears [constraint escapes](https://www.postgresql.org/docs/current/functions-matching.html#POSIX-CONSTRAINT-ESCAPES-TABLE) aren't working. Class-Shorthand escapes and Backreferences are working however.

Alkarex commented:
Thanks for the additional tests and feedback, @mtalexan 👍🏻 
Yes, the regex applied to other operators such as `author:XXX` will work identically.

Indeed, we do not have any feedback to the user when the search expression is wrong, for instance with wrong parentheses, invalid regex, etc. But we should add that.
And the whole parsing of the search query has grown from very simple rules to more complicated ones, and everything should be re-written with a proper parser. The question is how much should be done now, to avoid delaying the landing of the regex features (which we could brand as `beta` for now).

Adding an error feedback in particular should not be too difficult, but should still be left to a follow-up PR.

Alkarex committed:
Merge branch 'edge' into regex

mtalexan commented:
> Thanks for the additional tests and feedback, @mtalexan 👍🏻 Yes, the regex applied to other operators such as `author:XXX` will work identically.
> 
> Indeed, we do not have any feedback to the user when the search expression is wrong, for instance with wrong parentheses, invalid regex, etc. But we should add that. And the whole parsing of the search query has grown from very simple rules to more complicated ones, and everything should be re-written with a proper parser. The question is how much should be done now, to avoid delaying the landing of the regex features (which we could brand as `beta` for now).
> 
> Adding an error feedback in particular should not be too difficult, but should still be left to a follow-up PR.

I'm definitely on board with splitting the work up. Having access to regex parsing at all is a major feature that shouldn't be held up by reduced usability.

Alkarex committed:
Added author: and inurl: support, documentation

Alkarex committed:
author example

mtalexan commented:
Did you mean to include the `\b` (non-printing bell character) in the regex example in filtering Readmes? `intitle:/^Lo+l\b/i`

Alkarex commented:
Added `author:` and `inurl:` and some documentation (excluding the parentheses aspects for now). Additional tests welcome.

Example : `author:/^Alice Doe$/im` as we split multiple authors in different lines

Alkarex commented:
> Did you mean to include the `\b` (non-printing bell character) in the regex example in filtering Readmes? `intitle:/^Lo+l\b/i`

I meant `\b` like in [word boundary](https://www.regular-expressions.info/wordboundaries.html). Should be tested, by the way

Alkarex commented:
Hum, PostgreSQL does not support `\b`
https://www.postgresql.org/docs/current/functions-matching.html#POSIX-LIMITS-COMPATIBILITY

Alkarex committed:
Remove \b for now

mtalexan commented:
> > Did you mean to include the `\b` (non-printing bell character) in the regex example in filtering Readmes? `intitle:/^Lo+l\b/i`
> 
> I meant `\b` like in [word boundary](https://www.regular-expressions.info/wordboundaries.html). Should be tested, by the way

I thought that might be what you meant. PostgreSQL uses the TCL-style word boundaries instead (`\y`, `\Y`, `\m`, `\M`), not the Perl ones (`\b`). The PostgreSQL documentation calls them "constraint escapes", and they aren't working.

I reported the [word boundary escapes aren't working](https://github.com/FreshRSS/FreshRSS/pull/6706#issuecomment-2314227234), but I need to test the start and end of window matchers to see if they're also broken (the only other "constraint escapes" supported by PostgreSQL).

Alkarex commented on app/Models/Search.php:
```diff
@@ -226,9 +266,26 @@ public function getSearch(): ?array {
+	 * @return array<string>
+	 */
+	private static function sanitizeRegexes(array $regexes): array {
+		return array_filter($regexes, static fn(string $regex) => @preg_match($regex, '') !== false);
```
This does not work for PostgreSQL-specific regex syntax

Alkarex commented:
Right, the problem comes from the way I sanitize regex expressions, which does not work for PostgreSQL-specific regex syntax:

https://github.com/FreshRSS/FreshRSS/blob/afcdea0a426b9540d945d3246a702410959382b4/app/Models/Search.php#L286-L288

Edit: PostgreSQL example for later for myself: `intitle:/^Qui\M/`

Alkarex committed:
Disable regex sanitization for now

Alkarex commented:
I have disabled the regex sanitization for now. A question is whether it can be misused. Maybe not in a significant way (SQL ingestions should still not be possible).

Alkarex commented:
Another category of tests we should do relates to the proper handling of UTF-8. For instance, for the PHP regex engine, involving `\R` on multi-bytes strings (cf. https://github.com/FreshRSS/FreshRSS/issues/6236#issuecomment-2028492958). We should maybe have the `//u` modifier by default.

mtalexan commented:
> Another category of tests we should do relates to the proper handling of UTF-8. For instance, for the PHP regex engine, involving `\R` on multi-bytes strings (cf. [#6236 (comment)](https://github.com/FreshRSS/FreshRSS/issues/6236#issuecomment-2028492958)). We should maybe have the `//u` modifier by default.

Do you want me to run some tests on this, or are you saying it still needs to be enabled/implemented first?

Alkarex commented:
You are very welcome to give it a try

mtalexan commented on app/Models/Entry.php:
```diff
@@ -678,11 +698,21 @@ public function matches(FreshRSS_BooleanSearch $booleanSearch): bool {
 						$ok &= stripos($this->link, $url) !== false;
 					}
 				}
+				if ($ok && $filter->getIntitleRegex()) {
```
I think you meant this to be `getInurlRegex`

Alkarex replied:
Well spotted!

mtalexan commented on app/Models/EntryDAO.php:
```diff
@@ -960,7 +1013,7 @@ public static function sqlBooleanSearch(string $alias, FreshRSS_BooleanSearch $f
 			if ($filter->getSearch() !== null) {
```
Looks like you're missing the `getNotInurlRegex` case here

Alkarex replied:
Thanks! Fixed

mtalexan commented on app/Models/EntryDAO.php:
```diff
@@ -27,6 +27,34 @@ public static function sqlIgnoreConflict(string $sql): string {
+				// Case-sensitive matching
+				$matchType .= 'c';
+			}
+			$values[] = $matches['pattern'];
```
Looks like you haven't implemented multi-line matching for MySQL yet. Just adding this comment to track when it's completed.

Alkarex replied:
It looks like there might be a need for two different approaches for MySQL and for MariaDB.

MariaDB does not seem to support `REGEXP_LIKE` https://mariadb.com/kb/en/regexp/
but MariaDB has a new PCRE engine https://mariadb.com/kb/en/pcre/

Alkarex replied:
Implemented

mtalexan commented on app/Models/Entry.php:
```diff
@@ -636,21 +636,41 @@ public function matches(FreshRSS_BooleanSearch $booleanSearch): bool {
+						$ok &= preg_match($title, $this->title) === 0;
+					}
+				}
 				if ($ok && $filter->getTags()) {
```
It looks like it would be relatively easy to add tag regex support too. I know I've had cases where I wanted to match a list of tags and it's a a lot easier to do `#/\(tag1|tag2|tag3|tag4|tag5\)/i` than it is to do `#tag1 OR #tag2 OR #tag3 OR #tag4 OR #tag5`, especially if there are some shared sub-string between each of those tags.

Alkarex replied:
Done. Tests welcome

Alkarex replied:
P.S. In multiline

mtalexan commented on app/Models/EntryDAOPGSQL.php:
```diff
@@ -23,6 +23,32 @@ public static function sqlIgnoreConflict(string $sql): string {
+			$matchType = $matches['matchType'] ?? '';
+			if (str_contains($matchType, 'm')) {
+				// newline-sensitive matching
+				$matches['pattern'] = '(?m)' . $matches['pattern'];
```
It looks like metasyntax could be specified for Postgres, but only if you don't use `/../m` and include the `m` in the set of metasyntax options manually if you need it.  As a power user it's good to know, but I'm uncertain whether or not it might be worth mentioning in documentation somewhere.

mtalexan commented on app/Models/EntryDAOSQLite.php:
```diff
@@ -28,6 +28,27 @@ public static function sqlIgnoreConflict(string $sql): string {
+		// https://php.net/pdo.sqlitecreatefunction
+		// https://www.sqlite.org/lang_expr.html#the_like_glob_regexp_match_and_extract_operators
+		$this->pdo->sqliteCreateFunction('regexp',
+			function (string $pattern, string $text): bool {
```
It appears SQLite support isn't implemented yet. Just adding this comment so it's obvious when it is.

Alkarex replied:
Hum, SQLite is supposed to work already

mtalexan replied:
Sorry, that was from code inspection. I was looking for this block to do matching against the `im` like the others. I haven't migrated my database to check it yet though.  
This was just a comment to mark what was and wasn't included based on inspecting the code changes.

mtalexan commented:
Fun fact, `inurl:` is going to cause some difficulties.  Currently `inurl:` without regex doesn't work if you use quotes (https://github.com/FreshRSS/FreshRSS/issues/6750).  With regex now using the `/` as a regex "quoting" character, doing a URL match on the literal `/term/` is more difficult and existing patterns will break because `inurl:'/term/'` doesn't work, and the former literal `inurl:/term/` will now trigger a regex match instead.  Now the only way to do it is `inurl:/\/term\//`

mtalexan commented:
@Alkarex  I just realized something you mentioned before (emphasis mine):  

> [PHP preg_match](https://php.net/function.preg-match) for **_auto-mark-as-read and auto-favourite_**, as well as for searching with an SQLite database

This could present a pretty major problem if the regex engine and syntax for a filter run from the search box is different than the one used for auto-labeling and auto-marking-as-read. Currently the only way to test the filters for auto-labeling and auto-mark-as-read is to copy-paste it into the regular search box and see what comes up.  

I'm not familiar, is `preg_match` an option for all 3 database types and that's why it's used for auto-labeling and auto-mark-as-read rather than a database-specific syntax? Or is all the content being stored in an internal SQLite database during loading and auto-* matching before (possibly) being transferred to the alternative database? Or does searching on the SQLite database require PHP to do the filtering instead of the database, unlike other database types, and the auto-* matching occurs before the content reaches a database?

mtalexan commented:
If it's feasible, it seems like using the database-specific regex engine is almost always less powerful than using the PHP `preg_match`.  
Reading about it now, `preg_match` uses the PCRE library, which is literally the library for Perl-regex (but also including a few optional additions not implemented in Perl itself).  
Postgres has a modified form of the same (written from scratch), but with a more limited set of extended functions (e.g. `(?=...)`), and some modifications to a few of the escape codes (e.g. `\y` vs `\b`).  
MySQL only supports Extended-POSIX patterns with the addition of named character classes.  

Of all of them, PCRE is far and away more powerful, and avoids the non-standard Postgres syntax variations.  

---

Using the actual database engine for doing the search would appear to only really be useful for extremely massive datasets using an appropriately capable database so the engine itself can optimize the search.  To me it seems like it's more important to be able to test the auto-* matching against existing datasets than to have faster regex matching in manually entered/saved filters for huge datasets.  
If possible, I'd propose implementing regex thru the `preg_match` for all cases, then a follow up can look at providing an option to enable database-specific regex filtering (whether it's a toggle for the whole search filter, a global toggle, or a per-regex flag).

Alkarex committed:
Fix getInurlRegex

Alkarex committed:
Merge branch 'edge' into regex

Alkarex committed:
getNotInurlRegex

Alkarex committed:
Quotes for inurl:

Alkarex committed:
Fix test

Alkarex commented:
I would rather not embark on filtering client side (in PHP), except for SQLite which has a built-in way to do so.
Not only the performances would suffer, but the code would also be a lot more complicated with some parts to run server side and some parts to run client side

mtalexan commented:
> I would rather not embark on filtering client side (in PHP), except for SQLite which has a built-in way to do so. Not only the performances would suffer, but the code would also be a lot more complicated with some parts to run server side and some parts to run client side

Understandable. I only brought it up in case there was some architectural function I didn't see that would make it easy/feasible.  

---

Assuming you do the MariaDB split from MySQL, that would give 3 different regex engines: PCRE, a custom Tcl-style variant of Perl, and Extended POSIX + named character classes.  
You'll certainly want to make it big and bold in the filter document that labeling and auto-mark-read uses PHP pre-filtering regex engine while the search box uses the database regex engine and those syntaxes might not match. Putting together what you said then, the per-database(-ish) breakdown is:
- PHP (prefiltering) : PCRE
- SQLite (uses PHP): PCRE
- MariaDb : PCRE
- Postgres : custom Tcl-style variant of Perl syntax
- MySQL : Extended POSIX syntax + named character classes (no multiline support)

Notably, Extended POSIX + named character classes is a subset of the other types, and the custom Postgres format is similar to PCRE but more limited and with somw different identifiers (advanced use is often incompatible). It might be worth mentioning in the documemt that lays out the supported database types too since this could drive the database selection decision.

For me personally, it means I'll be switching my database to SQLite so I can have a consistent regex engine, but luckily there's no real harm for my use case since I only expect a few 10s of thousands of articles before purging starts kicking in. The database decision is primarily driven by that regex engine mismatch with Postgres though.

mtalexan commented:
To my future self:

> I'm not familiar, is preg_match an option for all 3 database types and that's why it's used for auto-labeling and auto-mark-as-read rather than a database-specific syntax? Or is all the content being stored in an internal SQLite database during loading and auto-* matching before (possibly) being transferred to the alternative database? Or does searching on the SQLite database require PHP to do the filtering instead of the database, unlike other database types, and the auto-* matching occurs before the content reaches a database?

It's using PHP to pre-filter incoming content as it's being added to the database so the associated labels and read state can be added at the same time. SQLite happens to use a PHP function you provide for it to do the filtering with when doing a database search.

Alkarex commented:
> luckily there's no real harm for my use case since I only expect a few 10s of thousands of articles

Note that SQLite's performance is surprisingly good, even with many articles, with or without regex

mtalexan commented:
@Alkarex  it appears the tag searching has the same issues as the `inurl:`, it doesn't work if quotes are included. https://github.com/FreshRSS/FreshRSS/issues/6761

Alkarex committed:
Merge branch 'edge' into regex

Alkarex committed:
Fix quoted tags + regex for tags
https://github.com/FreshRSS/FreshRSS/issues/6761

Alkarex commented:
Tags fixed (not tested). Tests welcome

Alkarex commented:
And regex added for tags

Alkarex committed:
Fix wrong regex detection

Alkarex committed:
Add MariaDB

Alkarex committed:
Fix logic

Alkarex committed:
Merge branch 'edge' into regex

Alkarex committed:
Increase requirements for MySQL and MariaDB
Check support for multiline mode in MySQL

Alkarex committed:
Remove sanitizeRegexes()

Alkarex committed:
Allow searching HTML code
Allow searching for instance `/<pre>/`
Fix https://github.com/FreshRSS/FreshRSS/issues/6775#issuecomment-2331769883

Alkarex committed:
Doc regex search HTML

Alkarex committed:
Fix Doctype

Alkarex committed:
Merge branch 'edge' into regex

mtalexan commented on docs/en/users/10_filter.md:
```diff
@@ -115,6 +117,31 @@ Finally, parentheses may be used to express more complex queries, with basic neg
+
+Example to search entries, which title starts with the *Lol* word, with any number of *o*: `intitle:/^Lo+l/i`
+
+As opposed to normal searches, HTML special characters are not escaped in regex searches, to allow searching HTML code, like: `/Hello <span>world<\/span>/`
```
This also has the effect that I have to use the HTML encoding of any character literals I'm looking for in the body as well, right? So `&amp;` and `&quot;` to find the post-rendered characters `&` and `"` for example.

Might it be worth doing this only if you have an `r` character on the regex?  Like `/Hello <span>world<\/span>/r`?

Alkarex replied:
Indeed, the search is supposed to match the HTML content verbatim. No need to extra encode anything though, just exactly at it appears in the original HTML document.

We could consider adding some modifiers, but let's see if there is a sufficient need first. Another modifier I am considering is an automated transformation from PCRE syntax (subset) to the other variants such as PostgreSQL.

mtalexan replied:
I think you might run into issues trying to translate between the regex versions. If you settle on the PCRE syntax there are a lot of advanced `(?...)` features that aren't present in Postgres.  Postgres also has explicit start and end word boundary markers, while PCRE has only a singular word boundary marker.  Then there's MySQL, which supports only a very small subset of regex features.  

I could see maybe trying to do something with the constraint escapes for Postgres, but that's about the only thing I can see that could possibly be translated and it would probably just confuse the situation with the databases having different regex formats.

Alkarex committed:
Merge branch 'edge' into pr/Alkarex/6706

Alkarex commented:
https://github.com/FreshRSS/FreshRSS/pull/6785

arseru commented:
Hi @Alkarex, I was trying to find issues/discussions where the regex filters were mentioned, and I ended up here. I have an issue where I'm trying to assign labels automatically to incoming feed items by using intitle regex filters, but it has some erratic behaviour that it's difficult for me to identify, and I was hoping you could give me a hint of what's happening.

So I tried the same as you did in your first comment for the FreshRSS releases feed:

![image](https://github.com/user-attachments/assets/7e396d6c-5f71-4870-a018-1f2bfe511e16)

If I use the filter `intitle:/\.0$/`, I get no results:

![image](https://github.com/user-attachments/assets/f79a4b09-6a8a-4d69-ae04-d5fdf95e2067)

Any idea why is this happening in my instance? I'm using the default [Docker deploy](https://github.com/FreshRSS/FreshRSS/tree/edge/Docker).

Alkarex commented:
> Any idea why is this happening in my instance? I'm using the default [Docker deploy](https://github.com/FreshRSS/FreshRSS/tree/edge/Docker).

What version of FreshRSS are you running? Try `freshrss/freshrss:edge` and make sure it was pulled recently

arseru commented:
> What version of FreshRSS are you running?

Version 1.24.3.

> Try `freshrss/freshrss:edge` and make sure it was pulled recently

Awesome! With this one (1.25.0-dev), it works as expected:

![image](https://github.com/user-attachments/assets/0d7a9643-e2a4-48f2-b6b6-662be19ca88b)

Problem solved then, thanks a lot for the quick reply :D

EDIT: duh, I just realized from your first comment that regex search is only available from FreshRSS 1.25.0+... sorry for not checking that before.

Alkarex commented:
@mtalexan Patch to allow parentheses in regex searches https://github.com/FreshRSS/FreshRSS/pull/6926
Tests welcome

Alkarex commented:
* https://github.com/FreshRSS/FreshRSS/pull/8141
